crates/rethnet_evm/src/blockchain/storage/contiguous.rs

```diff
@@ -0,0 +1,198 @@
+use std::sync::Arc;
+
+use rethnet_eth::{receipt::BlockReceipt, B256, U256};
+use revm::primitives::HashMap;
+
+use crate::{Block, LocalBlock};
+
+use super::InsertError;
+
+/// A storage solution for storing a Blockchain's blocks contiguously in-memory.
+#[derive(Clone, Default, Debug)]
+pub struct ContiguousBlockchainStorage<BlockT: Block + Clone + ?Sized> {
+    blocks: Vec<BlockT>,
+    hash_to_block: HashMap<B256, BlockT>,
+    total_difficulties: Vec<U256>,
+    transaction_hash_to_block: HashMap<B256, BlockT>,
+    transaction_hash_to_receipt: HashMap<B256, Arc<BlockReceipt>>,
+}
+
+impl<BlockT: Block + Clone> ContiguousBlockchainStorage<BlockT> {
+    /// Retrieves the instance's blocks.
+    pub fn blocks(&self) -> &[BlockT] {
+        &self.blocks
+    }
+
+    /// Retrieves a block by its hash.
+    pub fn block_by_hash(&self, hash: &B256) -> Option<&BlockT> {
+        self.hash_to_block.get(hash)
+    }
+
+    /// Retrieves the block that contains the transaction with the provided hash, if it exists.
+    pub fn block_by_transaction_hash(&self, transaction_hash: &B256) -> Option<&BlockT> {
+        self.transaction_hash_to_block.get(transaction_hash)
+    }
+
+    /// Retrieves the receipt of the transaction with the provided hash, if it exists.
+    pub fn receipt_by_transaction_hash(
+        &self,
+        transaction_hash: &B256,
+    ) -> Option<&Arc<BlockReceipt>> {
+        self.transaction_hash_to_receipt.get(transaction_hash)
+    }
+
+    /// Reverts to the block with the provided number, deleting all later blocks.
+    pub fn revert_to_block(&mut self, block_number: &U256) -> bool {
+        let block_number = usize::try_from(block_number)
+            .expect("No blocks with a number larger than usize::MAX are inserted");
+
+        let block_index = if let Some(first_block) = self.blocks.first() {
+            let first_block_number = usize::try_from(first_block.header().number)
+                .expect("No blocks with a number larger than usize::MAX are inserted");
```

Comment:
I think it'd be better to encode the invariant that the block number cannot exceed `usize::MAX` in the block type if it's not too much trouble. Also, I'm wondering about using `usize` vs `u64` here? I think the latter is safer for portability, since it's not inconceivable that block numbers will exceed 2^32 in the future.

Reply:
We intend to convert the type of block numbers from `U256` to `u64`. I've created an issue for it here: https://github.com/NomicFoundation/rethnet/issues/185